[guidelines] [writing_tests.md] Recommended tools (#14)
sadasant/azure-sdk-for-js: documentation/writing_tests.md

- [Introduction](#introduction)
- [Engineering setup](#engineering-setup)
- [Recommended tools](#recommended-tools)
  - [Mocha](#mocha)
  - [Chai](#chai)
  - [Rollup](#rollup)
  - [Karma](#karma)

>   - [The SDK's own test utilities](#the-sdk-s-own-test-utilities)

Author: I should put the recorder in this line.

```
rush add --dev -p mocha
```

To enhance the `mocha` experience, we're using the following packages:

- [`@types/mocha`](https://www.npmjs.com/package/@types/mocha), the type definitions for mocha.
- [`mocha-multi`](https://www.npmjs.com/package/mocha-multi), a way to get multiple reporters working with mocha.
- [`mocha-junit-reporter`](https://www.npmjs.com/package/mocha-junit-reporter), which produces JUnit-style XML test results.
- [`nyc`](https://www.npmjs.com/package/nyc) is [Istanbul](https://istanbul.js.org/)'s command line interface.
- [`esm`](https://www.npmjs.com/package/esm), a popular ECMAScript module loader.
- [`source-map-support`](https://www.npmjs.com/package/source-map-support), which provides [source map](https://github.com/mozilla/source-map) support for stack traces in node via the [V8 stack trace API](https://github.com/v8/v8/wiki/Stack-Trace-API).

A full `rush add` command that includes `mocha` and all of the previous dependencies follows:

```
rush add --dev -p mocha @types/mocha mocha-multi mocha-junit-reporter nyc esm source-map-support
```

We'll explore how we use these dependencies up next.

> #### Configuring Mocha

Author: I believe this section is missing a note on how to use Mocha from VSCode, but I don't ever do that, I use the command line. I can investigate, but perhaps somebody already knows.

Reviewer: If you're referring to debugging with VS Code, look at the debug config in storage, we have ways to debug the tests with and without rollup.

https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/storage/storage-blob/.vscode/launch.json#L25

We have the debug configs for samples too, but they might not work right now and might require little tweaks since the samples folder is modified structurally.

Maybe, you can keep a `To Do` for this.

## Recommended tools

Writing tests for JavaScript and TypeScript requires testing tools, such as a test framework, an assertion library and a way to bundle and run these tests on various environments. The JavaScript community has many overlapping tools that one could pick to fulfill any of these tasks. To ensure that the testing experience across the Azure SDK for JS/TS is consistent and reliable,

> we've picked the following external testing tools:

Reviewer 2: ```suggestion
Writing tests for JavaScript and TypeScript requires testing tools, such as a test framework, an assertion library, and a way to bundle and run these tests in various environments. The JavaScript community has many overlapping tools that one could pick to fulfill any of these tasks. To ensure that the testing experience across the Azure SDK for JS/TS is consistent and reliable, we've picked the following external testing tools:
```

Author: Thank you! Adding this 🌞

- [Mocha](https://www.npmjs.com/package/mocha), which offers us a well known and stable test framework for both NodeJS and browser tests.
- [Chai](https://www.npmjs.com/package/chai), a well known assertion library for Node and the browser.

> - [Rollup](https://www.npmjs.com/package/rollup), to bundle JavaScript for different environments, which helps us write a single TypeScript source and trust it will compile correctly for Node and the browser.

Reviewer 2: maybe some note that Rollup is good at creating efficient and small bundles for the browser out of ES6 modules?

Author: I just pushed a whole section on Rollup! Let me know if you like it.

- [Karma](https://www.npmjs.com/package/karma), which allows us to run our tests in multiple browsers.

> We've also come up with our own internal tools, which follow:

Reviewer 2: ```suggestion
We've also come up with our own internal tools:
```

- The [Recorder](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/recorder) is a tool that helps us run our live tests against static recordings obtained from a previous successful run, which aims to ensure that our code hasn't changed while benefitting from not having to reach out to real live services.

> Keep in mind that, over time, we might continue to expand these lists.

Reviewer 2: nit: not sure if this sentence is necessary

Author: I'll remove it!

Now, let's see how we're using the mentioned tools.

### Mocha

Mocha is a well known JavaScript test framework that runs in NodeJS and in the browser. If you're new to Mocha, we recommend you to dive into [Mocha's documentation](https://mochajs.org/).

> In this section, we'll explore how the Azure SDK for JS is using mocha, what versions we're using,how we're configuring it and any other recommendation on how to used this library.

Reviewer 2: ```suggestion
In this section, we'll explore how the Azure SDK for JS is using mocha, what versions we're using, how we're configuring it, and other recommendations on how to use this library.
```

Author: Thank you!

- [`esm`](https://www.npmjs.com/package/esm), a popular ECMAScript module loader.
- [`source-map-support`](https://www.npmjs.com/package/source-map-support), which provides [source map](https://github.com/mozilla/source-map) support for stack traces in node via the [V8 stack trace API](https://github.com/v8/v8/wiki/Stack-Trace-API).

A full `rush add` command that includes `mocha` and all of the previous dependencies follows:

```
rush add --dev -p mocha @types/mocha mocha-multi mocha-junit-reporter nyc esm source-map-support
```

We'll explore how we use these dependencies up next.

#### Configuring Mocha

Mocha is used by our test scripts in our `package.json` files. We typically invoke it with some common parameters, and some parameters that help us specify different test targets and different ways to debug the tests.

If we only consider the common parameters sent to mocha, we can see the following:

```
mocha --require esm --require source-map-support/register --reporter mocha-multi --reporter-options spec=-,mocha-junit-reporter=- --timeout 180000 --full-trace
```

Let's understand what's going on:

> - `--require esm` requires the `esm` ECMAScript module loader.

Reviewer 2: it feels a little weird to talk about this here if it's only for use with nyc and CI

Author: I can remove this. Should I do that? What's the best outcome in this case?

Author: `nyc` affects how we debug things though. See Harsha's response: https://github.com/sadasant/azure-sdk-for-js/pull/14/files#r392365613

Reviewer 2: maybe we start with the unit test case and then build on the nyc/integration case?

Author: That sounds good! I will capture this feedback later this week.

> - `--require source-map-support/register` requires the `source-map-support/register`, to support [source map](https://github.com/mozilla/source-map) for stack traces in node.

Reviewer 2: interestingly, not all projects use this option (and it's not in the template), so we should fix that soon

Author: How can we fix this? I am learning about our dependencies as I'm writing this.

- `--reporter mocha-multi --reporter-options spec=-,mocha-junit-reporter=-` which allows us to both use the `spec` reporter to output a hierarchical nested view of the test cases, and also produces the JUnit-style XML test results.
- `--timeout 180000`, which specifies the maximum time each single test case can take. More on that on the next section.
- `--full-trace`, which enables full stack traces, since Mocha by default shortens the stack traces.

That command by itself is still missing two things: the actual test files and a way to generate code coverage.
These two missing pieces can vary depending on what we're trying to test and how we're trying to debug the tests.

Code coverage can be added by placing `nyc` at the beginning of the line.
Keep in mind that `nyc` will **obscure the stack traces**, so it's preferable to
make separate `package.json` scripts, one for automated testing through CI, with `nyc`,

> and another one for developers running tests, without `nyc`.

Reviewer 2: This file doesn't consistently break long lines. I would honestly prefer that lines not be manually broken at inconsistent lengths and instead allow line-wrapping to do its magic.

Author: I started with manually broken lines, then I switched to line wrapping! But I missed some spots. I'll clean it up.

Then we have to point mocha to our test files. If you're **not** using `nyc`, you can point to the
bundled test file (bundled with rollup, which we will see later), typically at `dist-test/index.node.js`.

> If you are using `nyc`, point mocha to the files built by the TypeScript compiler, normally at `dist-esm/test/*.test.js`.

Reviewer 2: can you remind me why we run the test modules directly here instead of using the test bundle?

Author: I need to ping Harsha for this. I'll let him reply.

Reviewer: nyc code coverage was broken due to the bundling, In order to get the code coverage working properly, @jeremymeng changed the integration-test command to use the dist-esm files instead.

I think @sadasant pointed that the stack trace is getting messed up with the nyc code coverage when an error is thrown in the tests, making it harder to debug the sdk (or something in these lines). So, we eventually ended up having both versions of test commands.

Two ways of testing in our package.json's,
1. integration-test : with dist-esm files, and nyc code coverage
2. unit-test : with the rollup generated bundle, no code coverage

(Older integration-test command - with rollup generated bundle and nyc code coverage)

Reviewer 2: I guess there's no way to get nyc to import our sourcemaps?

Author: We tried for some time and failed. There could be a way. I wonder if @HarshaNalluru has more information.

Reviewer: I haven't done a deeper investigation or looked for a workaround to get nyc work with sourcemaps, my previous comment has all the info that I could recall.

    TODO: ASK ENGINEERING HOW TO GET THE MAX TIME PER TESTS

#### On the usage of before, beforeEach, after and afterEach

Mocha's `before` and `beforeEach` are methods that allow you to specify functions that should be executed either before all of the tests run or before each test runs, respectively. Similarly, `after` and `afterEach` exist so that developers can specify functions that should be executed after all tests run or after each test runs, respectively.

We recommend using `beforeEach` rather than `before`, just as much as we recommend using `afterEach` rather than `after`. The idea is that each one of your test cases should not depend on a state that is shared with other tests. Use `beforeEach` to execute tasks that will prepare the resources needed for each test to run cleanly, and `afterEach` to tear down or clean those settings before the next test runs.

We typically use `beforeEach` and `afterEach` to set up and tear down our test recorder. You can learn more about it in the section: [The Recorder](#the-recorder).

#### Other general recommendations

> Mocha has many interesting features. Here's a list of our general recommendations on how to use some of these features:

Reviewer 2: I would greatly appreciate a code snippet showing positive examples of how to write tests that are cleanly copy-pasteable

Author: I'm planning to have a whole section where I get into the details of how to write test cases 🌞 The title of that section is here: https://github.com/Azure/azure-sdk-for-js/pull/7784/files#diff-a43d448b275343b7d964b6e68f7cabf4R19

This PR is intended to have only recommendations about the tools we're using.

Having that in mind, would yous till like code snippets here? What kind of snippets? I can pick up and expand once I have a better idea of how to improve this part 🌞

Reviewer 2: a link to the other bit that talks about them would be fine

- **On async:**  
  Most of our test cases are asynchronous. Mocha allows you to write async test cases by calling `it` with an async function. Take this to your advantage. Use async functions on your test cases as much as possible. If all of your test cases use asynchronous functions, it will make them look and behave more consistently.
- **On arrow function expressions:**  
  Even though Mocha lets you write tests with arrow function expressions, as in `it("my test", async () => {})`, we recommend to use standard functions, as in `it("my test", async function () {})`, because standard functions have bindings to the execution context through the `this` keyword. Mocha's execution context is **necessary** to use the recorder, and it allows you to obtain the test name and other information that can be useful for your test cases. Mocha also discourages the use of arrow functions, which can be seen here: https://mochajs.org/#arrow-functions

### Chai

### Rollup

### Karma

### The Recorder

## package.json scripts

> Your package.json test scripts related to mocha should end up looking like this:

Reviewer 2: is it worth freezing these commands in a doc or should we link off to the template package.json?

Author: Linking to the template package.json is the way to go! This is heavily under development. Keep that in mind. Though, your feedback has been great so far! I really appreciate it. I will make sure to link to the template package.json 🌞

Author: I have added references to the template repo!

### Chai

Chai is an assertion library, similar to Node's built-in `assert`. While `assert` is a module that belongs to NodeJS's standard library, it was mainly designed to write tests for the NodeJS's core, and [it has been discouraged from being used as a dependency](https://github.com/nodejs/node/issues/4532). Chai is very popular and widely used for both NodeJS and the browser. You can read more about Chai through their main website: https://www.chaijs.com/

Chai makes testing much easier by providing an extensive list of assertions that can run against your code. This list of assertions can be seen in detail by going to their assertion style guide: <https://www.chaijs.com/guide/styles/>

Another important learning resource for Chai is: Chai Assertions for Promises <https://www.chaijs.com/plugins/chai-as-promised/>.

#### Our recommended Chai assertion style

The Azure SDK for JavaScript (and TypeScript) prefers Chai's BDD style with the `expect` interface. For example:

```ts
import * as chai from "chai";

let foo = "bar";
let beverages = { tea: [ 'chai', 'matcha', 'oolong' ] };

> expect(foo).to.be.a('string');

Reviewer 2: this surprises me, since most of the tests I've touched have used the `assert` style syntax, that seems more popular at MS. I don't mind expect syntax, but I think we should maybe do a straw poll in Teams about it.

Author: Sounds good. I picked one to start the conversation, but we should definitely ask the team.

Author: The poll leaned towards assert, so I'm pushing that!

```

You can see these dependencies at work in our [template project's package.json](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/template/template/package.json).

#### Configuring Rollup

Running `rollup` inside of a `package.json` script will automatically pick up the `rollup.config.js` file. What this file exports will define the configurations that `rollup` will use to make the bundle. We typically build for both Node and the browsers with the following script:

```json
  "build:nodebrowser": "rollup 2>&1",
```

Note that we add `2>&1` to hide the output of `rollup`.

Our `rollup.config.js` allows us to specify wether we want to do only the Node bundle, or only the browsers bundle, or both, by loading both configurations from a separate file, then picking either configuration, or both, depending on environment variables, `ONLY_NODE` to keep the NodeJS configuration, `ONLY_BROWSER` to keep the browser configuration, or neither to keep both. As follows:

```ts
import * as base from "./rollup.base.config";

const inputs = [];

if (!process.env.ONLY_BROWSER) {
  inputs.push(base.nodeConfig());
}

> // Disable this until we are ready to run rollup for the browser.

Reviewer 2: this comment feels weird?

Author: It's an oversight, I'll remove it!
